src/VoiceProtocol.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2005-2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MUMBLE_VOICEPROTOCOLTYPE_H
+#define MUMBLE_VOICEPROTOCOLTYPE_H
+
+#include "crypto/CipherType.h"
+#include "crypto/CryptStateFactory.h"
+
+#include <exception>
+#include <memory>
+#include <string>
+
+enum class VoiceTransportType { INVALID = 0, UDP = 1 };
+
+class VoiceProtocolException : public std::exception {};
+
+class VoiceProtocol {
+protected:
+	bool m_valid;
+	std::string m_protocolName;
+	explicit VoiceProtocol(const std::string &s);
+
+public:
+	VoiceTransportType m_transport;
+	virtual bool operator==(const VoiceProtocol &rhs) const;
+	virtual std::string toString() const;
+	static std::shared_ptr< VoiceProtocol > fromString(const std::string &s);
+	bool isValid() const;
+	VoiceProtocol();
+	virtual ~VoiceProtocol(){};
+};
+
+
+class UDPVoiceProtocol : public VoiceProtocol {
+private:
+	UDPVoiceProtocol(const std::string &s);
+	friend VoiceProtocol;
+
+public:
+	static std::shared_ptr< CryptStateFactory > m_cryptStateFactory;
```

Comment:
This static member is ugly, so is the initialization of this member. However, it has to know the cipher list from the `CryptStateFactory` in order to do the conversion.
I was thinking about storing the `CryptStateFactory` in Global and retrieve it by calling `Global::get()` but realized this is for Mumble only, not Murmur. In the end, I have to create this static member here and initialize them at the very beginning.

Reply:
You could simply create a getter-function that returns the factory and initializes it at first call into a function-static variable like this:

```cpp
CryptStateFactory &getFactory {
    static CryptStateFactory factory;

    return factory;
};
```

Reply:
Wow, this is great. Thanks for the advice!